### ISSUE TITLE
Ability to use Environment Flag with Branch Environments

### ISSUE THREAD

**Issue by homelinen** [github/octocatalog-diff#85]
From my testing and reading of https://github.com/github/octocatalog-diff/blob/master/doc/advanced-environments.md#preserving-the-environments it looks like it is not possible to use an environment other than 'production' when running octocatalog-diff.

The Puppet setup we have uses each branch as its own environment, and we currently manage 4 "production" like branches:

* production
* perftest
* international
* development

Our hiera hierarchy and a lot of code is dependant on the `environment` puppet variable so it is difficult to test our non-production branches.

So the ideal octocatalog-diff testing scenario would be to compare branch `feature/my_improvement` in environment 'perftest' against 'origin/production' in the 'perftest' environment. Optionally 'origin/perftest' in the perftest environment would also work.

This doesn't fit with the current requirement of `preserve-environments` having an `environments` folder in the repo, going by the example here: https://github.com/github/octocatalog-diff/blob/master/doc/advanced-environments.md#examples

The ability to specify an environment independent of the `preserve-environments` structure would be the best solution for my use case. I can see this is currently disabled here: https://github.com/github/octocatalog-diff/blob/0edf99bde6f205d3e428b91a28633b02e4b9704f/lib/octocatalog-diff/catalog-util/builddir.rb#L88 but haven't looked further into how that could be implemented.

I was going to look into the implementation of this for my use case but I wanted to raise a card to make sure there wasn't a strong reason not to implement something like this, or if it should be a separate set of command line options.

If I've totally misunderstood `--preserve-environments` I'd love to know how to achieve this use case.

**Comment by kpaulisse** [comment]
Hey @homelinen - so that I understand your request ... you want to be able to set `environment` to whatever you want (and to have this potentially be different in your "from" catalog and your "to" catalog), but you need :octocat:alog-diff to create a symlink from `environments/YOUR_ENVIRONMENT_HERE -> (puppet code)` instead of hard-coded `environments/production -> (puppet code)`?

I'll check out your PR as I noticed some of the tests were failing and I can probably help with that, but I want to be sure I understand the use case first.

**Comment by homelinen** [comment]
Yeah, that's exactly right @kpaulisse. Don't really need the environment to differ between to and from but absolutely no harm in that.

Thanks a lot for looking at the PR, I was getting a bit stuck on Friday! Will look over your suggestions properly later today.

**Comment by jamesandrews-iopp** [comment]
..can I just say "me too" to homelinen's suggestion.  We run a similar setup, using r10k

**Comment by jamesandrews-iopp** [comment]
Ok here's a hack that works with our setup.  I've only run --catalog-only so far

We have a git branch called "cloudlive" which is converted by r10k into a puppet environment on the puppetmaster.  The setup below works on a normal puppet agent node.  I have got a working ~/.octocatalog-diff.cfg.rb  the only thing that needed special attention was the puppetdb address, also I installed puppetdb-termini   on the agent node.

To make it run, set up a special directory with lots of copied stuff and symlinks :)

top level directory

[staffvx@succi puppete]$ ls -l
total 20
drwxr-xr-x. 3 staffvx ioppstaff 4096 Mar  6 10:11 environments
-rw-r--r--. 1 staffvx ioppstaff  543 Mar  6 14:19 hiera.yaml
drwxr-xr-x. 2 staffvx ioppstaff 4096 Mar  6 10:22 manifests
drwxr-xr-x. 7 staffvx ioppstaff 4096 Mar  6 10:14 modules

hiera.yml the normal, standard hiera file from /etc/puppetlabs/environment/heira.yml
manifests is an empty directory
modules is /etc/puppetlabs/code/modules copied from puppetmaster

environments has a cloudlive directory which contains
[staffvx@succi puppete]$ ls -l environments/cloudlive/
total 16
-rw-r--r--.  1 staffvx ioppstaff 1120 Mar  6 14:26 environment.conf
lrwxrwxrwx.  1 staffvx ioppstaff   36 Mar  6 09:42 hiera -> /home/staffvx/gitf/puppetmaster_np/hiera
lrwxrwxrwx.  1 staffvx ioppstaff   44 Mar  6 09:42 local_modules -> /home/staffvx/gitf/puppetmaster_np/local_modules
drwxr-xr-x.  2 staffvx ioppstaff 4096 Mar  6 09:43 manifests
drwxr-xr-x. 11 staffvx ioppstaff 4096 Mar  6 09:44 modules
-rw-r--r--.  1 staffvx ioppstaff  321 Mar  6 09:45 Puppetfile

heira and local_modules are symlinks to my git repos of puppet code
manifests contains site.pp copied from the puppetmaster
modules is copied from /etc/puppetlabs/code/environments/cloudlive/ on the puppetmaster

Puppetfile is copied from there too
environment.conf is hand modified
modulepath = ./modules:$basemodulepath:./local_modules:/home/staffvx/puppete/environments/cloudlive/local_modules:/home/staffvx/puppete/modules

I am not convinced that $basemodulepath doing what I hope


After doing this, commands such as

`octocatalog-diff  -n c1.webapplication \
 --catalog-only \
--bootstrapped-to-dir /home/staffvx/puppete \
--preserve-environments \
--to-environment cloudlive \
 --hiera-config /home/staffvx/puppete/hiera.yaml \
--output-format json`

Should work.
